app/src/test/java/org/mozilla/fenix/components/toolbar/CFRPresenterTest.kt

```diff
@@ -0,0 +1,324 @@
+        }
+
+        presenter.start()
+        assertNotNull(presenter.tcpCfrScope)
+
+        browserStore.dispatch(ContentAction.UpdateProgressAction(tab.id, 99)).joinBlocking()
+        browserStore.dispatch(ContentAction.UpdateProgressAction(tab.id, 100)).joinBlocking()
+        browserStore.dispatch(ContentAction.UpdateProgressAction(tab.id, 100)).joinBlocking()
+        browserStore.dispatch(ContentAction.UpdateProgressAction(tab.id, 100)).joinBlocking()
+        verify(exactly = 1) { presenter.showTcpCfr() }
+    }
+
+    @Test
+    fun `GIVEN the TCP CFR should not be shown WHEN the feature starts THEN don't observe the store for updates`() {
+        val presenter = CFRPresenter(
+            context = mockk(),
+            store = mockk(),
+            settings = mockk {
+                every { shouldShowTotalCookieProtectionCFR } returns false
+            },
+            toolbar = mockk(),
+        )
+
+        presenter.start()
+        assertNull(presenter.tcpCfrScope)
+    }
+
+    @Test
+    fun `GIVEN the store is observed for updates WHEN the presenter is stopped THEN stop observing the store`() {
+        val tcpScope: CoroutineScope = mockk {
+            every { cancel() } just Runs
+        }
+        val presenter = CFRPresenter(
+            context = mockk(),
+            store = mockk(),
+            settings = mockk(),
+            toolbar = mockk(),
+        )
+        presenter.tcpCfrScope = tcpScope
+
+        presenter.stop()
+
+        verify { tcpScope.cancel() }
+    }
+
+    @Test
+    fun `WHEN the TCP CFR is to be shown THEN instantiate a new one and remember to not show it again`() {
+        val settings: Settings = mockk(relaxed = true)
+        val presenter = CFRPresenter(
+            context = mockk(relaxed = true),
+            store = mockk(),
+            settings = settings,
+            toolbar = mockk(relaxed = true),
+        )
+
+        presenter.showTcpCfr()
+
+        verify { settings.shouldShowTotalCookieProtectionCFR = false }
+        assertNotNull(presenter.tcpCfr)
+    }
+
+    @Test
+    fun `WHEN the TCP CFR is instantiated THEN set the intended properties`() {
+        val context: Context = mockk {
+            every { getString(R.string.tcp_cfr_message) } returns "Test"
+        }
+        val anchor: View = mockk(relaxed = true)
+        val toolbar: BrowserToolbar = mockk {
+            every { findViewById<View>(R.id.mozac_browser_toolbar_security_indicator) } returns anchor
+        }
+        val settings: Settings = mockk(relaxed = true) {
+            every { toolbarPosition } returns ToolbarPosition.BOTTOM
+        }
+        val presenter = CFRPresenter(
+            context = context,
+            store = mockk(),
+            settings = settings,
+            toolbar = toolbar,
+        )
+
+        presenter.showTcpCfr()
+
+        verify { settings.shouldShowTotalCookieProtectionCFR = false }
+        val cfr = presenter.tcpCfr!!
```

Comment:
We have so few `!!`, may be there is a way to avoid using it? It looks like we are trying to avoid using it in the project in general!

Reply:
I agree that in production code we should avoid guaranteeing that the value is not null and if the property is nullable is better to use the elvis operator.
But in tests using the not null assertion operator is a test in itself: the value can be null but it shouldn't and I verify this by using `!!` instead of a separate `assertNotNull`.

Reply:
why we wouldn't use `assertNotNull` though? smth like this
```
presenter.tcpCfrPopup?.let {
            assertEquals("Test", it.text)
            assertEquals(anchor, it.anchor)
            assertEquals(CFRPopup.DEFAULT_WIDTH.dp, it.properties.popupWidth)
            assertEquals(CFRPopup.IndicatorDirection.DOWN, it.properties.indicatorDirection)
            assertTrue(it.properties.dismissOnBackPress)
            assertTrue(it.properties.dismissOnClickOutside)
            assertFalse(it.properties.overlapAnchor)
            assertEquals(CFRPopup.DEFAULT_INDICATOR_START_OFFSET.dp, it.properties.indicatorArrowStartOffset)
        }
       assertNotNull(presenter.tcpCfrPopup)
```

Reply:
Done.